File: solutions/150.Evaluate_Reverse_Polish_Notation/AC_stack_n.cpp
/*
*  Author:      illuz <iilluzen[at]gmail.com>
*  File:        AC_stack_n.cpp
*  Create Date: 2015-03-15 15:53:45
*  Descripton:  Using stack && atoi
*/

#include <bits/stdc++.h>

using namespace std;
const int N = 0;

class Solution {
public:
    int evalRPN(vector<string> &tokens) {
        stack<int> stk;
        string op = "+-*/";
        for (auto str : tokens) {
            auto pos = op.find(str);
            if (pos == string::npos) {
                stk.push(atoi(str.c_str()));
            } else {
                int a = stk.top();
                stk.pop();
                int b = stk.top();
                stk.pop();
                if (pos == 0)
                    stk.push(b + a);
                else if (pos == 1)
                    stk.push(b - a);
                else if (pos == 2)
                    stk.push(b * a);
                else if (pos == 3)
                    stk.push(b / a);
            }
        }
        return stk.top();
    }
};

int main() {

    return 0;
}

File: solutions/173.Binary_Search_Tree_Iterator/AC_stack_n.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
# Author:      illuz <iilluzen[at]gmail.com>
# File:        AC_stack_n.py
# Create Date: 2015-03-16 10:19:51
# Usage:       AC_stack_n.py 
# Descripton:  


# Definition for a  binary tree node
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class BSTIterator:
    # @param root, a binary search tree's root node
    def __init__(self, root):
        self.stack = []
        while root:
            self.stack.append(root)
            root = root.left

    # @return a boolean, whether we have a next smallest number
    def hasNext(self):
        return len(self.stack) != 0

    # @return an integer, the next smallest number
    def next(self):
        top = self.stack.pop()
        right = top.right
        while right:
            self.stack.append(right)
            right = right.left
        return top.val

# debug
root = TreeNode(0)
i, v = BSTIterator(root), []
while i.hasNext(): v.append(i.next())
